File: sample_billing_app_v3/backend/app/schemas/billing.py
"""Billing schemas."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.schemas.base import BaseSchema, TimestampSchema


class BillingWalletRead(BaseSchema):
    """Wallet balances for a user."""

    monthly_remaining: int = Field(..., ge=0)
    prepaid_balance: int = Field(..., ge=0)


class BillingSubscriptionRead(BaseSchema):
    """Subscription details."""

    status: str
    plan_name: str | None = None
    monthly_credits: int = 0
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    provider: str | None = None


class BillingCreditPack(BaseSchema):
    """Available credit pack."""

    credits: int
    price_usd: float


class TokenLedgerRead(TimestampSchema, BaseSchema):
    """Token usage ledger entry."""

    id: UUID
    model_name: str | None = None
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    total_tokens: int
    cost_credits: int
    overage_credits: int = 0


class BillingSummary(BaseSchema):
    """Billing summary response."""

    wallet: BillingWalletRead
    subscription: BillingSubscriptionRead | None
    credit_packs: list[BillingCreditPack]
    recent_ledger: list[TokenLedgerRead]


class CheckoutRequest(BaseSchema):
    """Checkout request."""

    kind: Literal["subscription", "credit_pack"]
    pack_credits: int | None = Field(default=None, description="Required for credit_pack")


class CheckoutResponse(BaseSchema):
    """Checkout response."""

    provider: str
    checkout_url: str

File: sample_billing_app_v2/frontend/src/types/billing.ts
export interface BillingWallet {
  monthly_remaining: number;
  prepaid_balance: number;
}

export interface BillingSubscription {
  status: string;
  plan_name?: string | null;
  monthly_credits: number;
  current_period_start?: string | null;
  current_period_end?: string | null;
  provider?: string | null;
}

export interface BillingCreditPack {
  credits: number;
  price_usd: number;
}

export interface TokenLedgerEntry {
  id: string;
  model_name?: string | null;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens: number;
  cost_credits: number;
  overage_credits: number;
  created_at: string;
}

export interface BillingSummary {
  wallet: BillingWallet;
  subscription: BillingSubscription | null;
  credit_packs: BillingCreditPack[];
  recent_ledger: TokenLedgerEntry[];
}

export type CheckoutKind = "subscription" | "credit_pack";

export interface CheckoutRequest {
  kind: CheckoutKind;
  pack_credits?: number;
}

export interface CheckoutResponse {
  provider: string;
  checkout_url: string;
}
